File: go/tree/levelOrder.go
package tree

func levelOrder(root *TreeNode) [][]int {
	var order [][]int
	if root == nil {
		return order
	}
	var q1 []*TreeNode
	q1 = append(q1, root)
	var q2 []*TreeNode
	var level []int
	for len(q1) != 0 {
		currentNode := q1[0]
		q1 = q1[1:]
		level = append(level, currentNode.Val)

		if currentNode.Left != nil {
			q2 = append(q2, currentNode.Left)
		}
		if currentNode.Right != nil {
			q2 = append(q2, currentNode.Right)
		}
		if len(q1) == 0 {
			q1 = q2
			q2 = []*TreeNode{}
			order = append(order, level)
			level = []int{}
		}
	}
	return order
}

func levelOrder2(root *TreeNode) [][]int {
	current := 0      // 当前层结点的数
	next := 0         // 下一层
	var order [][]int // 结果

	var q []*TreeNode // 队列
	if root != nil {
		q = append(q, root)
		current++
	}
	var level []int // 某一层结果
	for len(q) != 0 {
		currentNode := q[0]
		q = q[1:]
		current--
		level = append(level, currentNode.Val)

		if currentNode.Left != nil {
			q = append(q, currentNode.Left)
			next++
		}
		if currentNode.Right != nil {
			q = append(q, currentNode.Right)
			next++
		}
		if current == 0 {
			order = append(order, level)
			level = []int{}
			current = next
			next = 0
		}

	}
	return order
}

File: java/src/main/java/io/dure/tree/LevelOrder.java
package io.dure.tree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LevelOrder {

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> order = new ArrayList<>();
        Queue<TreeNode> q1 = new LinkedList<>();
        Queue<TreeNode> q2 = new LinkedList<>();
        List<Integer> level = new ArrayList<>();
        if (root != null) {
            q1.offer(root);
        }
        while (!q1.isEmpty()) {
            TreeNode currentNode = q1.poll();
            level.add(currentNode.val);
            if (currentNode.left != null) {
                q2.offer(currentNode.left);
            }
            if (currentNode.right != null) {
                q2.offer(currentNode.right);
            }
            if (q1.isEmpty()) {
                order.add(new ArrayList<>(level));
                level = new ArrayList<>();
                q1 = q2;
                q2 = new LinkedList<>();
            }
        }
        return order;
    }

    public List<List<Integer>> levelOrder2(TreeNode root) {
        List<List<Integer>> order = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        List<Integer> level = new ArrayList<>();
        int current = 0;
        int next = 0;
        if (root != null) {
            q.offer(root);
            current = 1;
        }
        while (!q.isEmpty()) {
            TreeNode currentNode = q.poll();
            current--;
            level.add(currentNode.val);
            if (currentNode.left != null) {
                q.offer(currentNode.left);
                next++;
            }
            if (currentNode.right != null) {
                q.offer(currentNode.right);
                next++;
            }
            if (current == 0) {
                order.add(new ArrayList<>(level));
                level = new ArrayList<>();
                current = next;
                next = 0;
            }
        }
        return order;
    }
}
